### PR TITLE
try of implementing cross-layer attention

### PR DESCRIPTION
**Pull request by Hrancheng**
Not sure whether this is correctly implemented, but a group of 4 attention layers(blocks) will use exactly the same K and V matrixes. We will work later to set this "4" to a user input argument.

### CODE REVIEW

**Comment by gkielian** [comment]
Looks good!

There are some merge conflicts from recent updates.

I'll create a pull request from my gkielian repo to Hrancheng:CLA_test for fixing these, will probably be arriving Thursday afternoon ish.

**Comment by gkielian** [comment]
Update on time estimate, will be starting a PR for the edits tomorrow, and seems I'll probably have the PR for review on Monday

File: model.py [hunk 1]
```diff
@@ -160,58 +161,72 @@ def __init__(self, config, fire_pos_enc=None):
...
-                gate_qx = Gating_q(x)
-                gate_q = torch.sigmoid(gate_qx)
-                gate_kv = torch.sigmoid(Gating_kv(gate_qx))
-                q = q * gate_q
-                k = k * gate_kv
-                v = v * gate_kv
+        ########CLA related modification#############
+        assert self.sharing_factor != 0
+        assert self.sharing_factor <= self.n_layer
+        if stored_k is not None and stored_v is not None and (block_count % self.sharing_factor) != 0:
```
[the remark this answers is not in the document]
Comment (gkielian) [reply model.py:171]: This is really interesting! : D

Looks like indeed the rotary embeddings can be skipped! (as they are already rotated)
This can saves a lot of processing in hardware : D 

Since the inputs were already checked for the kv, it's possible the gating might be able to be skipped as well...

This is worthy of exploration, add a todo for the above gate sharing feature.

Suggesting this to change to:
```python
if stored_k is not None and stored_v is not None and (block_count % self.sharing_factor) != 0:
        k = stored_k
        v = stored_v
        # TODO: Bring the gating into it's own method to prevent code duplication
        if self.gate and not self.use_cl_shared_gating:
            if self.n_kv_group == self.n_head:
                Gating = nn.Linear(self.n_embd, self.n_embd, bias=True, device=x.device)
                gate_ = torch.sigmoid(Gating(x))
                q = q * gate_
                k = k * gate_
                v = v * gate_
            else:
                # TODO: Test more methods to merge Attention Gates with GQA
                # TODO: Evaluate each method's ability to even out parameter sizes
                Gating_q = nn.Linear(self.n_embd, self.n_embd, bias=True, device=x.device)
                Gating_kv = nn.Linear(self.n_embd, self.kv_dim, bias=True, device=x.device)
                gate_qx = Gating_q(x)
                gate_q = torch.sigmoid(gate_qx)
                gate_kv = torch.sigmoid(Gating_kv(gate_qx))
                q = q * gate_q
                k = k * gate_kv
                v = v * gate_kv
 else:
        k = self.c_attn_k(x)
        v = self.c_attn_v(x)
        if self.rotary_emb_q is not None:
            q = self.rotary_emb_q(q)
            k = self.rotary_emb_k(k)

        if self.gate:
            if self.n_kv_group == self.n_head:
                Gating = nn.Linear(self.n_embd, self.n_embd, bias=True, device=x.device)
                gate_ = torch.sigmoid(Gating(x))
                q = q * gate_
                k = k * gate_
                v = v * gate_
            else:
                # TODO: Test more methods to merge Attention Gates with GQA
                # TODO: Evaluate each method's ability to even out parameter sizes
                Gating_q = nn.Linear(self.n_embd, self.n_embd, bias=True, device=x.device)
                Gating_kv = nn.Linear(self.n_embd, self.kv_dim, bias=True, device=x.device)
                gate_qx = Gating_q(x)
                gate_q = torch.sigmoid(gate_qx)
                gate_kv = torch.sigmoid(Gating_kv(gate_qx))
                q = q * gate_q
                k = k * gate_kv
                v = v * gate_kv
```

Also I think we need to leave windowed attention outside of the else block.

File: model.py [hunk 2]
```diff
@@ -160,58 +161,72 @@ def __init__(self, config, fire_pos_enc=None):
...
+            v = self.c_attn_v(x)
+        ########CLA related modification#############
+            if self.rotary_emb_q is not None:
+                q = self.rotary_emb_q(q)
+                k = self.rotary_emb_k(k)
 
+            if self.window_size is not None:
+                window_mask = torch.ones((1, 1, T, T), device=x.device)
+                window_mask = torch.triu(window_mask, diagonal=-self.window_size)
+                window_mask = self.bias[:,:,:T,:T] * window_mask
```
Comment (gkielian) [inline model.py:193]: I think we should move this outside of the else block, we should still mask with the same window if we are using the kv sharing

File: variations/softmax_variations.py [hunk 3]
```diff
@@ -202,13 +206,15 @@ def __init__(self, config, dim=-1):
 
         self.power = config.polymax_power
         self.divisor = config.polymax_divisor
+        self.inputs = []
+        self.outputs = []
```
Comment (gkielian) [inline variations/softmax_variations.py:210]: After experimenting with larger models, it seems these have a large overhead for processing time.

Add a new boolean setting to train.py and gpt_conf.py for `use_softmax_input_output_logging`, and then wrap lines with the self.inputs and self.outputs conditionally executed based on this setting.

File: variations/softmax_variations.py [hunk 4]
```diff
@@ -219,6 +225,7 @@ def forward(self, x):
         poly_piece = torch.where(x > 0, x**self.power + self.y_intercept, torch.tensor(0.0, device=x.device))
 
         # Combine sections
+
         result = (poly_piece + linear_piece + flat_piece)/self.divisor
 
         # divide by sequence length
```
Comment (gkielian) [inline variations/softmax_variations.py:231]: On line 235 before `return results` (github isn't letting me comment there so closest I can get).
add:

if self.use_softmax_input_output_logging:
    self.outputs = results

That way we can also monitor the results similarly to the strongermax

File: train.py [hunk 5]
```diff
@@ -559,6 +560,109 @@ def log_metrics_non_validation(self, loss_training, running_mfu, iter_num):
                 "mfu": running_mfu*100,
             })
 
+
```
Comment (gkielian) [inline train.py:563]: Since plotting statistics take a large overehead when testing, let's make the default args.statistic "None", and wrap the plot_statistics in the train.py:

```
if self.args.statistic is not None:
    self.plot_statistics()
```
And other areas as well where it is used

On another note, we should start thinking how to adapt the existing plot_statistics to also plot summary statistics of weights for each linear layer.

**Review by gkielian — CHANGES_REQUESTED** [review]
Looks great!

My review has some additional configuration settings and fields to add, afterwards should be ready to merge.

On a side note, if you can create slides about the approach and sections of the code modified to the team  -- specifically the for loop with the block and high level of what was modified ( gpt_conf configuration param, train.py argparse, model.py attention and block) -- could be a good chance to show an in depth modification process.

**Comment by Hrancheng** [comment]
> Looks great!
> 
> My review has some additional configuration settings and fields to add, afterwards should be ready to merge.
> 
> On a side note, if you can create slides about the approach and sections of the code modified to the team -- specifically the for loop with the block and high level of what was modified ( gpt_conf configuration param, train.py argparse, model.py attention and block) -- could be a good chance to show an in depth modification process.

Yes! I'll create slides along with codes modified next time to make this clear

**Comment by gkielian** [comment]
@Hrancheng reviving interest in this due to potential hardware intersection, let's discuss early next week : )

File: model.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (Hrancheng) [reply model.py:220]: Modified as above :)

File: model.py [hunk 6]
```diff
@@ -160,58 +161,72 @@ def __init__(self, config, fire_pos_enc=None):
...
+        else:
+            # print("not using stored k and v")
+            #x shape [5, 3, 128]
+            # print("block count: ", block_count)
+            # print("x shape: ", x.shape)
+            # print(x)
+            #
+            k = self.c_attn_k(x)
+            v = self.c_attn_v(x)
+        ########CLA related modification#############
```
[the remark this answers is not in the document]
Comment (Hrancheng) [reply model.py]: Sure, has adjusted this indent

File: model.py [hunk 7]
```diff
@@ -160,58 +161,72 @@ def __init__(self, config, fire_pos_enc=None):
...
+                    gate_qx = Gating_q(x)
+                    gate_q = torch.sigmoid(gate_qx)
+                    gate_kv = torch.sigmoid(Gating_kv(gate_qx))
+                    q = q * gate_q
+                    k = k * gate_kv
+                    v = v * gate_kv
         q = q.view(B, T, self.n_head, C // self.n_head).transpose(1, 2) # (B, n_h, T, hs)
-        k = k.view(B, T, self.n_kv_group, C // self.n_head).transpose(1, 2) # (B, n_kv, T, hs)
-        v = v.view(B, T, self.n_kv_group, C // self.n_head).transpose(1, 2) # (B, n_kv, T, hs)
+        reshaped_k = k.view(B, T, self.n_kv_group, C // self.n_head).transpose(1, 2) # (B, n_kv, T, hs)
```
[the remark this answers is not in the document]
Comment (Hrancheng) [reply model.py]: So I created store_k and store_v to store k and v before reshaping
        k_store = k
        v_store = v
        q = q.view(B, T, self.n_head, C // self.n_head).transpose(1, 2) # (B, n_h, T, hs)
        k = k.view(B, T, self.n_kv_group, C // self.n_head).transpose(1, 2) # (B, n_kv, T, hs)
        v = v.view(B, T, self.n_kv_group, C // self.n_head).transpose(1, 2) # (B, n_kv, T, hs)
Guess this will help clarify when we save non-reshaped k and v

File: model.py (hunk 2 above)
Comment (Hrancheng) [reply model.py]: Sure, has put this window mechanism outside when merging with upstream/master.

File: train.py (hunk 5 above)
Comment (Hrancheng) [reply train.py:729]: I'm working to put plot_statistics() in another folder as talked before

File: train.py [hunk 8]
```diff
@@ -71,6 +71,7 @@ def parse_args():
     model_group.add_argument('--block_size', default=256, type=int)
     model_group.add_argument('--n_layer', default=6, type=int)
     model_group.add_argument('--n_head', default=6, type=int)
+    model_group.add_argument('--sharing_factor', default=1, type=int)
```
Comment (gkielian) [inline train.py:74]: kv_crosslayer_sharing_factor

File: model.py [hunk 9]
```diff
@@ -466,23 +468,27 @@ def __init__(self, config, mlp=None, attn=None):
...
-                    x = self.ln_1(x + self.attn(x))
+                    attn_output, new_k, new_v = self.attn(x, stored_k, stored_v, block_count)
+                    x = self.ln_1(x + attn_output)
                     x = self.ln_2(x + self.mlp(x))
             else:
                 if self.use_parallel_mlp:
                     ln_1 = self.ln_1(x)
-                    x = x + self.attn(ln_1) + self.mlp(ln_1)
+                    attn_output, new_k, new_v = self.attn(x, stored_k, stored_v, block_count)
+                    x = x +attn_output + self.mlp(ln_1)
```
Comment (gkielian) [inline model.py:486]: add one space before attn

File: model.py [hunk 10]
```diff
@@ -212,7 +214,7 @@ def __init__(self, config, fire_pos_enc=None):
                                         .view(1, 1, config.block_size, config.block_size))
 
 
-    def forward(self, x):
+    def forward(self, x, stored_k=None, stored_v=None, block_count=0, kv_cache_table=None):
```
Comment (gkielian) [inline model.py:217]: note kv_cache_table

File: model.py
Comment (gkielian) [inline model.py:1]: kv sharing factor should be utilized somewhere maybe in gpt2 forward pass

**Review by gkielian — CHANGES_REQUESTED** [review]
